torch/sparse/semi_structured.py

```diff
@@ -0,0 +1,318 @@
+from collections import namedtuple
+import warnings
+
+import torch
+
+
+__all__ = [
+    "to_sparse_semi_structured",
+    "SparseSemiStructuredTensor",
+]
+
+CUSPARSELT_CONFIG = namedtuple("CUSPARELT_CONFIG", "compression_factor n min_size")
+DTYPE_TO_CUSPARSELT_CONFIG = {
+    torch.float32 : CUSPARSELT_CONFIG(9, 1, 8),
+    torch.float16 : CUSPARSELT_CONFIG(9, 2, 16),
+    torch.bfloat16 : CUSPARSELT_CONFIG(9, 2, 16),
+    torch.int8: CUSPARSELT_CONFIG(10, 4, 32),
+}
+
+class SparseSemiStructuredTensor(torch.Tensor):
```

Comment:
nit: We should be careful about not giving an overly generic name that we might want to need later.
Having this file contain everything related to semi structured sparsity sounds good. But the subclass that implements only 2:4 sparsity and that won't be able to be extended for other semi-structured dtypes doesn't sound great. Can we make this TwoFourSparseTensor or something along those line to be very specific?

Reply:
The subclass actually implements 1:2 sparsity support for fp32 and 2:4 sparsity support for int8, fp16, and bf16, so I don't think TwoFourSparseTensor is the best name. 
Semi structured refers to the fact that half (semi) of the elements are kept, which is the same for all dtypes.

I'll update the docstring to be make that more clear.

Reply:
Ho the "Semi" part was surprising.
I would defer to @cpuhrsch for a more informed opinion on this but we should make sure this is open to adding more as they come along.

Reply:
@albanD - I'm also not sure what the best naming is. "semi" does mean "half" as a prefix and for all these patterns (1:2, 2:4, 4:8) we always have to set half the values to zero. It's structured because we're splitting everything into blocks. It also doesn't collide with the names of other sparse layouts I can think of.